modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/pagemem/SpeedBasedCleanPagesProtectionThrottle.java

```diff
@@ -0,0 +1,500 @@
+
+        if (ratio >= 0)
+            return ratio;
+
+        return pageMemory.getDirtyPagesRatio();
+    }
+
+    /**
+     * @return Returns {@link #speedForMarkAll}.
+     */
+    public long getLastEstimatedSpeedForMarkAll() {
+        return speedForMarkAll;
+    }
+
+    /**
+     * @return Speed average checkpoint write speed. Current and 3 past checkpoints used. Pages/second.
+     */
+    public long getCpWriteSpeed() {
+        return cpWriteSpeed.getOpsPerSecondAtNow();
+    }
+
+    /***/
+    int threadIdsCount() {
+        return threadIds.size();
+    }
+
+    /**
+     * @return Counter for fsynced checkpoint pages.
+     */
+    int cpSyncedPages() {
+        AtomicInteger syncedPagesCntr = cpProgress.apply().syncedPagesCounter();
+
+        return syncedPagesCntr == null ? 0 : syncedPagesCntr.get();
+    }
+
+    /**
+     * @return Number of pages in current checkpoint.
+     */
+    int cpTotalPages() {
+        return cpProgress.apply().currentCheckpointPagesCount();
+    }
+
+    /**
+     * @return number of evicted pages.
+     */
+    int cpEvictedPages() {
+        AtomicInteger evictedPagesCntr = cpProgress.apply().evictedPagesCounter();
+
+        return evictedPagesCntr == null ? 0 : evictedPagesCntr.get();
+    }
+
+    /**
+     * @param baseSpeed   speed to slow down.
+     * @return sleep time in nanoseconds.
+     */
+    long calcDelayTime(long baseSpeed) {
+        return calcDelayTime(baseSpeed, threadIdsCount(), 1);
+    }
+
+    /**
+     * @param baseSpeed   speed to slow down.
+     * @param nThreads    operating threads.
+     * @param factor      how much it is needed to slowdown base speed. 1 means delay to get exact base speed.
+     * @return sleep time in nanoseconds.
+     */
+    private long calcDelayTime(long baseSpeed, int nThreads, int factor) {
+        if (factor <= 0)
+            throw new IllegalStateException("Coefficient should be positive");
+
+        if (baseSpeed <= 0)
+            return 0;
+
+        long updTimeNsForOnePage = TimeUnit.SECONDS.toNanos(1) * nThreads / (baseSpeed);
+
+        return factor * updTimeNsForOnePage;
+    }
+
+    /**
+     * @param cpWrittenPages  current counter of written pages.
+     * @param dirtyPagesRatio current percent of dirty pages.
+     */
+    private void detectCpPagesWriteStart(int cpWrittenPages, double dirtyPagesRatio) {
+        if (cpWrittenPages > 0 && lastObservedWritten.compareAndSet(0, cpWrittenPages)) {
+            double newMinRatio = dirtyPagesRatio;
+
+            if (newMinRatio < MIN_RATIO_NO_THROTTLE)
+                newMinRatio = MIN_RATIO_NO_THROTTLE;
+
+            if (newMinRatio > 1)
+                newMinRatio = 1;
+
+            //for slow cp is completed now, drop previous dirty page percent
+            initDirtyRatioAtCpBegin = newMinRatio;
+        }
+    }
+
+    /**
+     * Resets the throttle to its initial state (for example, in the beginning of a checkpoint).
+     */
+    void initialize() {
```

Comment:
Naming here isn't perfect: `initialize` implies to be called once, but this method is called over and over again on the same object.

How about `reset` as a name for this method?

Reply:
I agree. Renamed to `reset()`.